mix/MixClient.cpp

```diff
@@ -140,92 +198,155 @@ bytes MixClient::call(Secret _secret, u256 _value, Address _dest, bytes const& _
 	ReadGuard l(x_state);
-	return m_state.balance(_a);
+	return asOf(_block).balance(_a);
 }
 
 u256 MixClient::countAt(Address _a, int _block) const
 {
-	validateBlock(_block);
 	ReadGuard l(x_state);
-	return m_state.transactionsFrom(_a);
+	return asOf(_block).transactionsFrom(_a);
 }
 
 u256 MixClient::stateAt(Address _a, u256 _l, int _block) const
 {
-	validateBlock(_block);
 	ReadGuard l(x_state);
-	return m_state.storage(_a, _l);
+	return asOf(_block).storage(_a, _l);
 }
 
 bytes MixClient::codeAt(Address _a, int _block) const
 {
-	validateBlock(_block);
 	ReadGuard l(x_state);
-	return m_state.code(_a);
+	return asOf(_block).code(_a);
 }
 
 std::map<u256, u256> MixClient::storageAt(Address _a, int _block) const
 {
-	validateBlock(_block);
 	ReadGuard l(x_state);
-	return m_state.storage(_a);
+	return asOf(_block).storage(_a);
 }
 
 eth::LocalisedLogEntries MixClient::logs(unsigned _watchId) const
 {
-	(void)_watchId;
-	return LocalisedLogEntries();
+	Guard l(m_filterLock);
+	h256 h = m_watches.at(_watchId).id;
+	auto filterIter = m_filters.find(h);
+	if (filterIter != m_filters.end())
+		return logs(filterIter->second.filter);
+	return eth::LocalisedLogEntries();
 }
 
-eth::LocalisedLogEntries MixClient::logs(eth::LogFilter const& _filter) const
+eth::LocalisedLogEntries MixClient::logs(eth::LogFilter const& _f) const
 {
-	(void)_filter;
-	return LocalisedLogEntries();
+	LocalisedLogEntries ret;
+	unsigned lastBlock = m_blocks.size() - 1; //last block contains pending transactions
+	unsigned block = std::min<unsigned>(lastBlock, (unsigned)_f.latest());
+	unsigned end = std::min(lastBlock, std::min(block, (unsigned)_f.earliest()));
+	for (; ret.size() != _f.max() && block != end; block--)
+	{
+		bool pendingBlock = (block == lastBlock);
+		if (pendingBlock || _f.matches(m_blocks[block].info.logBloom))
+			for (ExecutionResult const& t: m_blocks[block].transactions)
+				if (pendingBlock || _f.matches(t.receipt.bloom()))
+				{
+					LogEntries logEntries = _f.matches(t.receipt);
+					if (logEntries.size())
+					{
+						for (unsigned entry = _f.skip(); entry < logEntries.size() && ret.size() != _f.max(); ++entry)
+							ret.insert(ret.begin(), LocalisedLogEntry(logEntries[entry], block));
+					}
+				}
+	}
+	return ret;
+}
+
+unsigned MixClient::installWatch(h256 _h)
+{
+	unsigned ret;
+	{
+		Guard l(m_filterLock);
+		ret = m_watches.size() ? m_watches.rbegin()->first + 1 : 0;
+		m_watches[ret] = ClientWatch(_h);
+	}
+	auto ch = logs(ret);
+	if (ch.empty())
+		ch.push_back(eth::InitialChange);
+	{
+		Guard l(m_filterLock);
+		swap(m_watches[ret].changes, ch);
+	}
+	return ret;
 }
 
-unsigned MixClient::installWatch(eth::LogFilter const& _filter)
+unsigned MixClient::installWatch(eth::LogFilter const& _f)
 {
-	(void)_filter;
-	BOOST_THROW_EXCEPTION(InterfaceNotSupported("dev::eth::Interface::installWatch"));
+	h256 h = _f.sha3();
+	{
+		Guard l(m_filterLock);
+		if (!m_filters.count(h))
+			m_filters.insert(std::make_pair(h, _f));
```

Comment:
insert does not insert if the key already exists, so the check above is redundant.